application/data/standardisers/ethnicity_classification_finder.py

```diff
@@ -0,0 +1,287 @@
+class EthnicityClassificationFinder:
+    """
+    EthnicityClassificationFinder is our advanced standardiser used by ChartBuilder2 and TableBuilder2
+
+    EthnicityClassificationFinder uses the observation that government ethnicity data uses certain defined
+    classifications and these determine how charts and tables should be displayed. See the dashboards for examples
+
+    EthnicityClassificationFinder first converts raw entry data to standard labels from the Race Disparity Audit
+    Then it searches our classification library for possible matches from known classifications
+    """
+
+    def __init__(self, ethnicity_standardiser, ethnicity_classification_collection):
+        self.standardiser = ethnicity_standardiser
+        self.classification_collection = ethnicity_classification_collection
+
+    def find_classifications(self, raw_ethnicities):
+        valid_classifications = self.__get_valid_classifications(raw_ethnicities)
+
+        classification_data = [
+            classification.get_outputs(raw_ethnicities, self.standardiser) for classification in valid_classifications
+        ]
+        custom_data = EthnicityClassification.get_custom_data_outputs(raw_ethnicities)
+
+        all_output_data = classification_data + [custom_data]
+
+        return all_output_data
+
+    @staticmethod
+    def find_classifications_for_default_finder(raw_ethnicities):
+        custom_data = EthnicityClassification.get_custom_data_outputs(raw_ethnicities)
+        return [custom_data]
+
+    def __get_valid_classifications(self, raw_ethnicities):
+        return self.classification_collection.get_valid_classifications(raw_ethnicities, self.standardiser)
+
+
+class EthnicityStandardiser:
+    def __init__(self, ethnicity_map=None):
+        if ethnicity_map:
+            self.ethnicity_map = ethnicity_map
+        else:
+            self.ethnicity_map = {}
+
+    def __len__(self):
+        return len(self.ethnicity_map)
+
+    def add_conversion(self, raw_ethnicity, standard_ethnicity):
+        lookup_value = self.simplify_key(raw_ethnicity)
+        self.ethnicity_map[lookup_value] = standard_ethnicity
+
+    @classmethod
+    def simplify_key(cls, key_value):
+        return key_value.strip().lower()
+
+    def standardise(self, raw_ethnicity):
+        lookup_value = self.simplify_key(raw_ethnicity)
+        if lookup_value in self.ethnicity_map:
+            return self.ethnicity_map[lookup_value]
+        else:
+            return raw_ethnicity
+
+    def standardise_all(self, raw_ethnicity_list):
+        return [self.standardise(raw_ethnicity) for raw_ethnicity in raw_ethnicity_list]
```

Comment:
It looks like you use `raw_ethnicities` in most other places - be cool to use it here as well for the sake of consistency.

Reply:
done